src/app/components/guards/auth.tsx

```diff
@@ -1,22 +1,48 @@
-import useHasStateRehydrated from '@hooks/stores/useHasRehydrated';
+import { ReactNode, useEffect, useState } from 'react';
+import { useNavigate } from 'react-router-dom';
+import useSecretKey from '@hooks/useSecretKey';
 import useWalletSelector from '@hooks/useWalletSelector';
-import { Navigate } from 'react-router-dom';
 
-function AuthGuard({ children }: React.PropsWithChildren) {
-  const { encryptedSeed, seedPhrase } = useWalletSelector();
-  const hydrated = useHasStateRehydrated();
+interface AuthGuardProps {
+  children: ReactNode;
+}
 
-  if (hydrated && encryptedSeed && !seedPhrase) {
-    return <Navigate to="/login" />;
-  }
+function AuthGuard({ children }: AuthGuardProps) {
+  const navigate = useNavigate();
+  const { masterPubKey, encryptedSeed } = useWalletSelector();
+  const { getSeed, hasSeed } = useSecretKey();
+  const [authTested, setAuthTested] = useState(false);
 
-  if (hydrated && !encryptedSeed) {
-    return <Navigate to="/landing" />;
-  }
+  const restoreSession = async () => {
+    if (encryptedSeed) {
+      navigate('/login');
+      setAuthTested(true);
+      return;
+    }
+    const hasSeedPhrase = await hasSeed();
+    if (!hasSeedPhrase || !masterPubKey) {
+      navigate('/landing');
+      setAuthTested(true);
+      return;
+    }
+    try {
+      await getSeed();
+      setAuthTested(true);
+    } catch (error) {
+      navigate('/login');
+    }
+    setAuthTested(true);
+  };
 
-  // fragment is required here
-  // eslint-disable-next-line react/jsx-no-useless-fragment
-  return <>{children}</>;
+  useEffect(() => {
+    restoreSession();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  if (!authTested) {
+    return null;
+  }
+  return children;
```

Comment:
the fragment was required to fix eslint errors, as children can be an array of ReactNodes

Reply:
Maybe we should add a comment?
```suggestion
  // fragment is required here because without it, the router thinks there could be more than 1 child node
  // eslint-disable-next-line react/jsx-no-useless-fragment
  return children;
```

Reply:
resolved